api/envoy/config/listener/v3/listener.proto

```diff
@@ -119,6 +119,40 @@ message Listener {
   // :ref:`FAQ entry <faq_how_to_setup_sni>`.
   repeated FilterChain filter_chains = 3;
 
+  // A list of filter chain matching predicate extension type URLs specifying
+  // the order of selection for the most specific filter chain. At each step,
+  // the predicate is evaluated for each filter chain. If the predicate returns
+  // false, the filter chain is eliminated from subsequent consideration. If
+  // the predicate returns true, but there is a more specific match, it is also
+  // eliminated.
+  //
+  // For example, for SNI ``www.example.com`` the filter chain with the
+  // predicate ```*.example.com``` remains and the filter chain with the
+  // predicate ```*.com``` is eliminated, assuming there is no filter chain
+  // with the exact SNI match. In general, the order of specificity is domain
+  // specific as defined by the predicate extension.
+  //
+  // In case, there is more than one filter chain candidates remaining after
+  // the process completes, the first of the filter chains in the order of
+  // their declaration is selected.
+  //
+  // If left unspecified, the following default matching order is applied:
+  // 1. :ref:`DestinationPort <envoy_v3_api_msg_config.listener.v3.FilterChainMatch.DestinationPort>`.
+  // 2. :ref:`DestinationIP <envoy_v3_api_msg_config.listener.v3.FilterChainMatch.DestinationIP>`.
+  // 3. :ref:`SNI <envoy_v3_api_msg_config.listener.v3.FilterChainMatch.SNI>`.
+  // 4. :ref:`TransportProtocol <envoy_v3_api_msg_config.listener.v3.FilterChainMatch.TransportProtocol>`.
+  // 5. :ref:`ApplicationProtocol <envoy_v3_api_msg_config.listener.v3.FilterChainMatch.ApplicationProtocol>`.
+  // 6. :ref:`DirectSourceIP <envoy_v3_api_msg_config.listener.v3.FilterChainMatch.DirectSourceIP>`.
+  // 7. :ref:`SourceType <envoy_v3_api_msg_config.listener.v3.FilterChainMatch.SourceType>`.
+  // 8. :ref:`SourceIP <envoy_v3_api_msg_config.listener.v3.FilterChainMatch.SourceIP>`.
+  // 9. :ref:`SourcePort <envoy_v3_api_msg_config.listener.v3.FilterChainMatch.SourcePort>`.
+  //
+  // Filter chain match conditions that are not in the matching order list are
```

Comment:
I'm debating whether we should perform a linear check for predicates that are not in this list instead. That seems more intuitive.

Reply:
Sorry what do you mean by this exactly?

Reply:
Imagine someone adds a new filter chain match predicate using the metadata. If we don't assume this order is complete, then there will be multiple matching filter chains left by default for metadata matching.